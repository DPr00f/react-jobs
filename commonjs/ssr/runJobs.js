'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = runJobs;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactTreeWalker = require('react-tree-walker');

var _reactTreeWalker2 = _interopRequireDefault(_reactTreeWalker);

var _utils = require('../utils');

var _ServerProvider = require('./ServerProvider');

var _ServerProvider2 = _interopRequireDefault(_ServerProvider);

var _createRunJobsExecContext = require('./createRunJobsExecContext');

var _createRunJobsExecContext2 = _interopRequireDefault(_createRunJobsExecContext);

var _constants = require('./constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getJobs(rootElement, rootContext) {
  var fetchRoot = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;

  var jobs = [];

  (0, _reactTreeWalker2.default)(rootElement, function (element, instance, context) {
    var skipRoot = !fetchRoot && element === rootElement;
    if (instance && typeof instance.getExecutingJob === 'function' && !skipRoot) {
      var job = instance.getExecutingJob();
      if ((0, _utils.isPromise)(job)) {
        jobs.push({ job: job, element: element, context: context });

        // Tell walkTree to not recurse inside this component;  we will
        // wait for the query to execute before attempting it.
        return false;
      }
    }

    return undefined;
  }, rootContext);

  return jobs;
}

function runJobs(rootElement) {
  var rootContext = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  var isRoot = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;

  var processingElement = void 0;
  if (isRoot) {
    var runJobsExecContext = (0, _createRunJobsExecContext2.default)();
    rootContext.runJobsExecContext = runJobsExecContext; // eslint-disable-line no-param-reassign
    processingElement = _react2.default.createElement(
      _ServerProvider2.default,
      { runJobsExecContext: runJobsExecContext },
      rootElement
    );
  } else {
    processingElement = rootElement;
  }

  var resolveResult = function resolveResult() {
    return {
      appWithJobs: processingElement,
      state: rootContext.runJobsExecContext.getState(),
      STATE_IDENTIFIER: _constants.STATE_IDENTIFIER
    };
  };

  var runningJobs = getJobs(processingElement, rootContext, isRoot)
  // Map over the jobs and then...
  .map(function (_ref) {
    var job = _ref.job,
        element = _ref.element,
        context = _ref.context;
    return (
      // ... make sure they will continue executing jobs for their Children
      // when they are complete.
      job.then(function () {
        return runJobs(element, context,
        // We've just grabbed the job for element so don't try and get it again
        false);
      })
    );
  });

  return runningJobs.length > 0 ? Promise.all(runningJobs)
  // Swallow errors.
  .catch(function () {
    return undefined;
  }).then(function () {
    return resolveResult();
  }) : Promise.resolve(resolveResult());
}