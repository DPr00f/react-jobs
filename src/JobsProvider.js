/* @flow */

import { Children, Component, Element, PropTypes } from 'react';
import type { ProviderContext } from './sharedTypes';

type React$Element = Element<*>;

type Props = {
  children?: React$Element,
};

let currentjobID = 0;
const jobResults = {};

const context : ProviderContext = {
  reactJobs: {
    nextJobID: () => {
      currentjobID += 1;
      return currentjobID;
    },
    registerJobResults: (jobID: number, results: any) => {
      jobResults[jobID] = results;
    },
    getJobResults: (jobID: number) => jobResults[jobID],
  },
};

class JobsProvider extends Component {
  props: Props;

  getChildContext() {
    return context;
  }

  render() {
    return Children.only(this.props.children);
  }
}

JobsProvider.childContextTypes = {
  reactJobs: PropTypes.object.isRequired,
};

export default JobsProvider;
