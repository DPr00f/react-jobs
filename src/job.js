/* @flow */

import React, { Component, PropTypes } from 'react';
import { getDisplayName, isPromise } from './utils';
import type { ProviderContext } from './sharedTypes';

type Work = (props : Object) => any;

type State = {
  resolved?: boolean,
  inProgress: boolean,
  result?: any,
  error?: any,
  executingWork? : Promise<any>,
};

type Props = Object;

export default function job(work : Work) {
  if (typeof work !== 'function') {
    throw new Error('You must provide a function to a react-jobs job declaration.');
  }

  // Oh I love you closures!
  let jobID = null;

  return function WrapComponentWithJob(WrappedComponent : Function) {
    class ComponentWithJob extends Component {
      props: Props;
      state: State;

      constructor(props : Props, context : ?ProviderContext) {
        super(props);
        if (context && context.reactJobs) {
          if (!jobID) {
            jobID = context.reactJobs.nextJobID();
          } else {
            const jobResults = context.reactJobs.getJobResults(jobID);
            if (jobResults) {
              this.state = Object.assign(
                {},
                jobResults,
                { resolved: true },
              );
            }
          }
        }
        if (!this.state) {
          this.state = { inProgress: false };
        }
      }

      componentWillMount() {
        if (this.state.resolved) {
          return;
        }

        const context : ProviderContext = this.context;
        const x = work(this.props);

        if (isPromise(x)) {
          const executingWork = x
            .then((result) => { this.setState({ inProgress: false, result }); return result; })
            .catch((error) => {
              console.warn('Job failed:\n', error); // eslint-disable-line no-console
              this.setState({ inProgress: false, error });
            })
            .then(() => {
              if (jobID) {
                context.reactJobs.registerJobResults(jobID, this.state);
              }
            });

          // Asynchronous result.
          this.setState({ inProgress: true, executingWork });
        } else {
          // Synchronous result.
          this.setState({ result: x });
        }
      }

      getExecutingJob() {
        if (!this.state.executingWork) {
          return undefined;
        }
        return this.state.executingWork;
      }

      render() {
        const { inProgress, result, error } = this.state;
        const jobProp = { inProgress, result, error };
        return <WrappedComponent {...this.props} job={jobProp} />;
      }
    }
    ComponentWithJob.displayName = `${getDisplayName(WrappedComponent)}WithJob`;
    ComponentWithJob.contextTypes = { reactJobs: PropTypes.object };
    return ComponentWithJob;
  };
}
