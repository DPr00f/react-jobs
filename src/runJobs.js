/* @flow */

import { Children, Element } from 'react';
import { isPromise } from './utils';

type React$Element = Element<*>;
type Context = { [key: string]: any; };
type ElementVisitor =
  (element: React$Element, instance: ?Function, context: Context) => boolean | void;
type ElementJob = {
  job: Promise<any>,
  element: React$Element,
  context: Context,
}

// Recurse an React Element tree, running visitor on each element.
// If visitor returns `false`, don't call the element's render function
//   or recurse into its child elements
export function walkTree(
  element: React$Element,
  context: Context,
  visitor: ElementVisitor,
) {
  const Component = element.type;

  // a stateless functional component or a class
  if (typeof Component === 'function') {
    const props = Object.assign({}, Component.defaultProps, element.props);
    let childContext = context;
    let child;

    // Are we are a react class?
    // http://bit.ly/2j9Ifk3
    const isReactClassComponent = Component.prototype &&
      (Component.prototype.isReactComponent || Component.prototype.isPureReactComponent);

    if (isReactClassComponent) {
      const instance = new Component(props, context);
      // In case the user doesn't pass these to super in the constructor
      instance.props = instance.props || props;
      instance.context = instance.context || context;

      // Override setState to just change the state, not queue up an update.
      //   (we can't do the default React thing as we aren't mounted "properly"
      //   however, we don't need to re-render as well only support setState in
      //   componentWillMount, which happens *before* render).
      instance.setState = (newState) => {
        instance.state = Object.assign({}, instance.state, newState);
      };

      if (instance.componentWillMount) {
        instance.componentWillMount();
      }

      if (instance.getChildContext) {
        childContext = Object.assign({}, context, instance.getChildContext());
      }

      if (visitor(element, instance, context) === false) {
        return;
      }

      child = instance.render();
    } else { // just a stateless functional
      if (visitor(element, null, context) === false) {
        return;
      }

      child = Component(props, context);
    }

    if (child) {
      walkTree(child, childContext, visitor);
    }
  } else { // a basic string or dom element, just get children
    if (visitor(element, null, context) === false) {
      return;
    }

    if (element.props && element.props.children) {
      Children.forEach(element.props.children, (child: any) => {
        if (child) {
          walkTree(child, context, visitor);
        }
      });
    }
  }
}

function getJobs(
  rootElement : React$Element,
  rootContext : Context,
  fetchRoot : boolean = true,
) : ElementJob[] {
  const jobs = [];

  walkTree(rootElement, rootContext, (element, instance, context) => {
    const skipRoot = !fetchRoot && (element === rootElement);
    if (instance && typeof instance.getExecutingJob === 'function' && !skipRoot) {
      const job = instance.getExecutingJob();
      if (isPromise(job)) {
        jobs.push({ job, element, context });

        // Tell walkTree to not recurse inside this component;  we will
        // wait for the query to execute before attempting it.
        return false;
      }
    }

    return undefined;
  });

  return jobs;
}

export default function runJobs(
  rootElement : React$Element,
  rootContext : Context = {},
  fetchRoot : boolean = true,
) {
  const jobs = getJobs(rootElement, rootContext, fetchRoot);

  // no queries found, nothing to do
  if (!jobs.length) {
    return Promise.resolve(null);
  }

  // wait on each query that we found, re-rendering the subtree when it's done
  const mappedJobs = jobs.map(({ job, element, context }) =>
    // we've just grabbed the query for element so don't try and get it again
    job.then(() => runJobs(element, context, false)),
  );

  return Promise.all(mappedJobs).then(() => null);
}
