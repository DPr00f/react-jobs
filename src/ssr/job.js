/* @flow */
// @see https://github.com/facebook/react/issues/1137

import React, { PropTypes } from 'react';
import { getDisplayName } from '../utils';
import clientOnlyJob from '../job';
import type { ReactJobsClientContext, ReactJobsServerContext } from './types';

type Context = {
  reactJobsClient: ReactJobsClientContext,
  reactJobsServer?: ReactJobsServerContext,
};

const getInitialState = (context, jobID) => {
  const { reactJobsClient, reactJobsServer } = context;

  if (reactJobsServer) {
    // Running on the server. If the jobs have been executed by the runJobs
    // helper then we will get back a result here.
    const jobState = reactJobsServer.getJobState(jobID);
    if (jobState) {
      return jobState;
    }
  } else {
    // Otherwise we have to be on the browser. Check if there is any
    // rehydration state available for this job.
    const ssrRehydrateState = reactJobsClient.popJobRehydrationForSRR(jobID);
    if (ssrRehydrateState) {
      return ssrRehydrateState;
    }
  }

  return undefined;
};

const job = (work : any) => (Component : Function) => {
  let jobIDHandle = null;

  function withSSRBehaviour(WrappedComponent) {
    const ComponentWithJobID = (props : Object, context : Context) => {
      const { reactJobsClient, reactJobsServer } = context;

      // Every SSR job should have the client attached in order to provide
      // the identifiers for the components.
      if (!reactJobsClient) {
        throw new Error('You are using an "ssr/job" without having wrapped your component render with the "runJobs" or "rehydrateJobs" helpers.  Please refer to the docs for more details.');
      }

      // Establish a unique identifier for this job instance.
      const getJobID = () => {
        if (!jobIDHandle) {
          jobIDHandle = reactJobsClient.nextJobID();
        }
        return jobIDHandle;
      };
      const jobID = getJobID();

      // Determine if we have an state from server side processing or
      // client side rehdration.
      const initialState = getInitialState(context, jobID);

      // When on a server we want to call back and register the result of the
      // work so that the server can hydrate the app appropriately as well
      // as provide the initial state to be sent to the browser.
      const onJobProcessed = reactJobsServer
        ? jobsState => reactJobsServer.registerJobState(jobID, jobsState)
        : null;

      return (
        <WrappedComponent
          {...props}
          jobInitState={initialState}
          onJobProcessed={onJobProcessed}
        />
      );
    };
    ComponentWithJobID.displayName = `${getDisplayName(WrappedComponent)}WithJobID`;
    ComponentWithJobID.contextTypes = {
      reactJobsClient: PropTypes.object.isRequired,
      reactJobsServer: PropTypes.object,
    };
    return ComponentWithJobID;
  }

  // We wrap the standard job implementation with our SSR behaviour.
  return withSSRBehaviour(clientOnlyJob(work)(Component));
};

export default job;
