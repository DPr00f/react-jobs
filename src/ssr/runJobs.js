/* @flow */

// eslint-disable-next-line import/no-extraneous-dependencies
import React, { Element } from 'react'
// eslint-disable-next-line import/no-extraneous-dependencies
import reactTreeWalker from 'react-tree-walker'
import { isPromise } from '../utils'
import ServerProvider from './ServerProvider'
import createRunJobsExecContext from './createRunJobsExecContext'
import { STATE_IDENTIFIER } from './constants'
import type { RehydrateState } from './types'

type React$Element = Element<*>;
type Context = { [key: string]: any };
type ElementJob = {
  job: Promise<any>,
  element: React$Element,
  context: Context,
};
type RunJobsResult = {
  appWithJobs: React$Element,
  state: RehydrateState,
  STATE_IDENTIFIER: string,
};

function getJobs(
  rootElement: React$Element,
  rootContext: Object,
  fetchRoot: boolean = true,
): ElementJob[] {
  const jobs = []

  reactTreeWalker(
    rootElement,
    (element, instance, context) => {
      const skipRoot = !fetchRoot && element === rootElement
      if (
        instance && typeof instance.getExecutingJob === 'function' && !skipRoot
      ) {
        const job = instance.getExecutingJob()
        if (isPromise(job)) {
          jobs.push({ job, element: instance, context })

          // Tell walkTree to not recurse inside this component;  we will
          // wait for the query to execute before attempting it.
          return false
        }
      }

      return undefined
    },
    rootContext,
  )

  return jobs
}

export default function runJobs(
  rootElement: Element<any>,
  rootContext: Object = {},
  isRoot: boolean = true,
): Promise<RunJobsResult> {
  let processingElement
  if (isRoot) {
    const runJobsExecContext = createRunJobsExecContext()
    rootContext.runJobsExecContext = runJobsExecContext // eslint-disable-line no-param-reassign
    processingElement = (
      <ServerProvider runJobsExecContext={runJobsExecContext}>
        {rootElement}
      </ServerProvider>
    )
  } else {
    processingElement = rootElement
  }

  const resolveResult = () => ({
    appWithJobs: processingElement,
    state: rootContext.runJobsExecContext.getState(),
    STATE_IDENTIFIER,
  })

  const runningJobs = getJobs(processingElement, rootContext, isRoot)
    // Map over the jobs and then...
    .map(({ job, element, context }) =>
      // ... make sure they will continue executing jobs for their Children
      // when they are complete.
      job.then(() =>
        runJobs(
          element,
          context,
          // We've just grabbed the job for element so don't try and get it again
          false,
        )))

  return runningJobs.length > 0
    ? Promise.all(runningJobs)
        // Swallow errors.
        .catch(() => undefined)
        .then(() => resolveResult())
    : Promise.resolve(resolveResult())
}
