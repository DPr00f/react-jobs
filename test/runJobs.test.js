/* @flow */

import React from 'react';
import { mount } from 'enzyme';
import JobsProvider from '../src/JobsProvider';
import job from '../src/job';
import { Foo, resolveAfter } from './__helpers__';

// Under test.
import runJobs from '../src/runJobs';

const workTime = 10; // ms

describe('runJobs()', () => {
  it('should return a Promise', () => {
    const actual = runJobs(<div>foo</div>);
    expect(actual.then).toBeTruthy();
  });

  describe('integration', () => {
    it('should render simple job with success', () => {
      const FooWithJob = job(() => resolveAfter(workTime, 'bob'))(Foo);
      const app = (
        <JobsProvider>
          <FooWithJob />
        </JobsProvider>
      );
      return runJobs(app).then(() => {
        expect(mount(app)).toMatchSnapshot();
      });
    });
  });
});
