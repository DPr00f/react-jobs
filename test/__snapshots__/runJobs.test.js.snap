exports[`runJobs() integration should render simple job with success 1`] = `
<JobsProvider>
  <FooWithJob>
    <Foo
      job={
        Object {
          "error": undefined,
          "inProgress": false,
          "result": "bob",
        }
      }>
      <div>
        bob
      </div>
    </Foo>
  </FooWithJob>
</JobsProvider>
`;
